app/agent/activities/utils.ts

```diff
@@ -0,0 +1,25 @@
+import { Context } from "@temporalio/activity";
+
+export const withActivityHeartbeat = <T>(
+  options: {
+    /** Default: 5000 */
+    intervalMs?: number;
+  },
+  fn: (...args: any[]) => Promise<T>,
+): Promise<T> => {
+  return new Promise((resolve, reject) => {
+    const interval = setInterval(() => {
+      Context.current().heartbeat();
+    }, options.intervalMs ?? 5000);
+
+    fn()
+      .then((result) => {
+        clearInterval(interval);
+        resolve(result);
+      })
+      .catch((err) => {
+        clearInterval(interval);
+        reject(err);
+      });
```

Comment:
```suggestion
    return fn()
      .finally(() => {
        clearInterval(interval);
      });
```

Reply:
I did this intentionally - clearInterval is called before the promise ends to ensure that it's done before any other code is executed.